mars/oscar/__init__.py

```diff
@@ -12,11 +12,66 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import asyncio
 from .api import actor_ref, create_actor, has_actor, destroy_actor
-from .core import Actor, ActorRef
+from .core import _Actor, ActorRef
 from .context import get_context
 from .errors import ActorNotExist, ActorAlreadyExist
 from .utils import create_actor_ref
 
 from .backends import mars
 del mars
+
+
+_actor_implementation = dict()
+
+
+def register_actor_implementation(actor_cls, impl_cls):
+    _actor_implementation[id(actor_cls)] = impl_cls
+
+
+def unregister_actor_implementation(actor_cls):
+    try:
+        del _actor_implementation[id(actor_cls)]
+    except KeyError:
+        pass
+
+
+# Cython has a bug about inspecting async function: https://github.com/cython/cython/issues/2273
```

Comment:
We do not need to use ``asyncio.iscoroutinefunction``. Checking return value is enough in most scenarios. Thus please reconsider moving this code into ``__init__``.

Reply:
But ray uses `asyncio.iscoroutinefunction` to determine whether to run async or not.

Reply:
I'd rather add a util to enable ``asyncio.iscoroutinefunction`` by adding a util function like:

```python
# add in a pure .py file
def aio_coro_function(func):
    async def dummy():
        pass
    func._is_coroutine = getattr(dummy, '_is_coroutine')
```

Reply:
Then we need to use this `iscoroutinefunction` in ray or monkey patch ray.

Reply:
A tested diff can be https://github.com/mars-project/mars/compare/refactor/oscar...wjsi:refactor/oscar

Reply:
> Then we need to use this `iscoroutinefunction` in ray or monkey patch ray.

Sorry deleted a response by mistake. In my implementation above, no changes needed for ``iscoroutinefunction``.

Reply:
A bit hacky. I will try it, thanks.

Reply:
Ray is using the `inspect. iscoroutinefunction` not the `asyncio. iscoroutinefunction`: https://github.com/ray-project/ray/blob/master/python/ray/_raylet.pyx#L421

Reply:
> Ray is using the `inspect.iscoroutinefunction` not the `asyncio.iscoroutinefunction`: https://github.com/ray-project/ray/blob/master/python/ray/_raylet.pyx#L421
> 
> I can create a PR for ray to use `asyncio.iscoroutinefunction` instead.

Now in #1925 , the Actor class has been moved to `api.py` which is not a cython file any more.

Reply:
Thanks.